File: old-python-backend/src/game/battle/combat.py
"""战斗伤害计算与攻击执行。"""

from __future__ import annotations

import math
import random
from typing import Any

from game.battle.core import _to_float, _clamp
from game.battle.entity import _refresh_entity_stats
from game.battle.skills import (
    _skill_has_damage,
    _skill_requires_effect_change,
    _skill_would_apply_effect,
    _apply_skill_effects,
    _evaluate_skill_conditions,
)
from game.battle.rewards import (
    _record_enemy_death,
    _add_pending_skill_exp,
    _finish_wave_if_needed,
    _handle_player_down,
)


def _calc_damage_result(attacker: dict[str, Any], defender: dict[str, Any], skill: dict[str, Any]) -> dict[str, Any]:
    damage_cfg = skill.get("damage") or {}
    damage_type = str(damage_cfg.get("type") or "physical").lower()
    flat = _to_float(damage_cfg.get("flat"), 0.0)
    mult = _to_float(damage_cfg.get("multiplier"), 1.0)

    attacker_acc = max(1e-6, _to_float(attacker.get("accuracy"), 1.0))
    defender_evade = max(1e-6, _to_float(defender.get("evade"), 1.0))
    evade_possibility = 1.0 / (1.0 + (_clamp(attacker_acc / defender_evade, 0.0, 10.0) ** 2))
    evaded = random.random() < evade_possibility

    rand_ratio = 0.9 + random.random() * 0.2

    if damage_type == "magic":
        attack_power = max(0.0, flat + _to_float(attacker.get("magic_power"), 0.0) * mult)
        raw_damage = (
            rand_ratio
            * attack_power
            / max(1e-6, 1.0 + _to_float(defender.get("magic_instance"), 0.0))
            * (1.0 + _to_float(attacker.get("final_damage_multiplier"), 0.0))
            / max(1e-6, 1.0 + _to_float(defender.get("final_damage_reduce"), 0.0))
        )
        raw_damage = max(0.0, raw_damage)
        return {
            "damage": 0.0 if evaded else raw_damage,
            "raw_damage": raw_damage,
            "damage_type": "magic",
            "evaded": evaded,
            "blocked": False,
            "blocked_reduction": 0.0,
        }

    attack_power = max(0.0, flat + _to_float(attacker.get("physical_power"), 0.0) * mult)
    defense = max(0.0, _to_float(defender.get("defense"), 0.0))
    original_damage = (
        rand_ratio
        * (attack_power ** 2)
        / max(1e-6, attack_power + defense)
        * (1.0 + _to_float(attacker.get("final_damage_multiplier"), 0.0))
        / max(1e-6, 1.0 + _to_float(defender.get("final_damage_reduce"), 0.0))
    )
    original_damage = max(0.0, original_damage)

    if evaded:
        return {
            "damage": 0.0,
            "raw_damage": original_damage,
            "damage_type": "physical",
            "evaded": True,
            "blocked": False,
            "blocked_reduction": 0.0,
        }

    block = max(0.0, _to_float(defender.get("block"), 0.0))
    block_pos_mult = max(0.0, _to_float(defender.get("block_possibility_multiplier"), 0.0))
    block_possibility = 1.0 - 100.0 / (100.0 + block) / max(1e-6, 1.0 + block_pos_mult)
    block_possibility = _clamp(block_possibility, 0.0, 1.0)
    blocked = random.random() < block_possibility
    blocked_reduction = 0.0
    if blocked:
        block_rate = max(1.0, _to_float(defender.get("block_rate"), 1.0))
        blocked_damage = max(0.0, original_damage / block_rate)
        blocked_reduction = max(0.0, original_damage - blocked_damage)
        original_damage = blocked_damage

    if random.random() < _clamp(_to_float(attacker.get("critical"), 0.0), 0.0, 1.0):
        original_damage *= max(1.0, _to_float(attacker.get("critical_rate"), 1.0))
    original_damage = max(0.0, original_damage)

    return {
        "damage": original_damage,
        "raw_damage": original_damage + blocked_reduction,
        "damage_type": "physical",
        "evaded": False,
        "blocked": blocked,
        "blocked_reduction": blocked_reduction,
    }


def _choose_player_skill(session: dict[str, Any], target_enemy: dict[str, Any] | None) -> dict[str, Any]:
    available_skills = session["player"]["skills"]
    basic_skill_id = session["player"]["basic_skill_id"]
    now = session["time"]
    plan = session["player"]["skill_plan"]
    alive_enemies = [e for e in session["enemies"] if e["alive"]]

    for entry in plan:
        skill_id = entry["skill_id"]
        skill = available_skills.get(skill_id)
        if not skill:
            continue
        if _to_float(session["player"]["cooldowns"].get(skill_id), 0.0) > now:
            continue
        mp_cost = max(0.0, _to_float(skill.get("mp_cost"), 0.0))
        sp_cost = max(0.0, _to_float(skill.get("sp_cost"), 0.0))
        if _to_float(session["player"].get("mp"), 0.0) < mp_cost:
            continue
        if _to_float(session["player"].get("sp"), 0.0) < sp_cost:
            continue
        if not _evaluate_skill_conditions(entry.get("condition"), session["player"], target_enemy, alive_enemies):
            continue
        if (
            not _skill_has_damage(skill)
            and _skill_requires_effect_change(skill)
            and not _skill_would_apply_effect(
            skill,
            session["player"],
            target_enemy,
            now,
        )
        ):
            continue
        return skill

    return available_skills.get(basic_skill_id) or {
        "id": "__basic_attack__",
        "name": "基础攻击",
        "damage": {"type": "physical", "flat": 0.0, "multiplier": 1.0},
        "cooldown": 0.0,
        "cast_time": _to_float(session["player"]["stats"].get("attack_interval"), 2.0),
    }


def _choose_enemy_skill(
    session: dict[str, Any],
    enemy: dict[str, Any],
    target_player: dict[str, Any] | None,
) -> dict[str, Any]:
    available_skills = enemy.get("skills") or {}
    basic_skill_id = str(enemy.get("basic_skill_id") or "__enemy_basic_attack__")
    now = _to_float(session.get("time"), 0.0)
    plan = enemy.get("skill_plan") or []
    alive_allies = [e for e in session.get("enemies") or [] if isinstance(e, dict) and e.get("alive")]

    for entry in plan:
        skill_id = str(entry.get("skill_id") or "")
        if not skill_id:
            continue
        skill = available_skills.get(skill_id)
        if not skill:
            continue
        if _to_float((enemy.get("cooldowns") or {}).get(skill_id), 0.0) > now:
            continue

        mp_cost = max(0.0, _to_float(skill.get("mp_cost"), 0.0))
        sp_cost = max(0.0, _to_float(skill.get("sp_cost"), 0.0))
        if _to_float(enemy.get("mp"), 0.0) < mp_cost:
            continue
        if _to_float(enemy.get("sp"), 0.0) < sp_cost:
            continue
        if not _evaluate_skill_conditions(entry.get("condition"), enemy, target_player, alive_allies):
            continue
        if (
            not _skill_has_damage(skill)
            and _skill_requires_effect_change(skill)
            and not _skill_would_apply_effect(
            skill,
            enemy,
            target_player,
            now,
        )
        ):
            continue
        return skill

    fallback = available_skills.get(basic_skill_id)
    if isinstance(fallback, dict):
        return fallback
    return {
        "id": "__enemy_basic_attack__",
        "name": "基础攻击",
        "damage": {"type": str(enemy.get("basic_damage_type") or "physical"), "flat": 0.0, "multiplier": 1.0},
        "cooldown": 0.0,
        "cast_time": _to_float((enemy.get("stats") or {}).get("attack_interval"), 2.0),
        "effects": [],
    }


def _process_player_attack(session: dict[str, Any], logs: list[dict[str, Any]]) -> None:
    alive_enemies = [e for e in session["enemies"] if e["alive"]]
    if not alive_enemies:
        return

    player = session["player"]
    target = alive_enemies[0]
    skill = _choose_player_skill(session, target)
    mp_cost = max(0.0, _to_float(skill.get("mp_cost"), 0.0))
    sp_cost = max(0.0, _to_float(skill.get("sp_cost"), 0.0))
    if mp_cost > 0:
        player["mp"] = _clamp(_to_float(player.get("mp"), 0.0) - mp_cost, 0.0, _to_float(player.get("max_mp"), 0.0))
        _add_pending_skill_exp(session, "intelligence", 5.0 * mp_cost)
    if sp_cost > 0:
        player["sp"] = _clamp(_to_float(player.get("sp"), 0.0) - sp_cost, 0.0, _to_float(player.get("max_sp"), 0.0))
        _add_pending_skill_exp(session, "stamina", 5.0 * sp_cost)

    applied_effects = _apply_skill_effects(session, skill, player, target)

    damage_result = {
        "damage": 0.0,
        "evaded": False,
        "blocked": False,
        "damage_type": "physical",
    }
    dmg = 0.0
    if _skill_has_damage(skill):
        damage_result = _calc_damage_result(player["stats"], target["stats"], skill)
        dmg = _to_float(damage_result.get("damage"), 0.0)
        target["hp"] = _clamp(target["hp"] - dmg, 0.0, target["max_hp"])

        if dmg > 0:
            damage_type = str(damage_result.get("damage_type") or "physical").lower()
            if damage_type == "magic":
                _add_pending_skill_exp(session, "magic", 1.0 * dmg)
            else:
                physical_style = str(skill.get("physical_style") or "melee").lower()
                if physical_style == "ranged":
                    _add_pending_skill_exp(session, "ranging", 0.5 * dmg)
                    _add_pending_skill_exp(session, "strength", 0.5 * dmg)
                else:
                    _add_pending_skill_exp(session, "strength", 0.8 * dmg)
                    _add_pending_skill_exp(session, "ranging", 0.2 * dmg)

    cast_time = max(0.1, _to_float(skill.get("cast_time"), _to_float(player["stats"].get("attack_interval"), 2.0)))
    cooldown = max(0.0, _to_float(skill.get("cooldown"), 0.0))

    player["next_ready_time"] = session["time"] + cast_time
    player["last_action_duration"] = cast_time
    player["last_skill_id"] = skill.get("id")
    player["last_skill_name"] = skill.get("name") or skill.get("id")
    player["cooldowns"][skill["id"]] = session["time"] + cooldown

    logs.append(
        {
            "type": "player_attack",
            "skill_id": skill.get("id"),
            "skill_name": skill.get("name") or skill.get("id"),
            "target_id": target["enemy_id"],
            "target_instance_id": target["instance_id"],
            "damage": round(dmg, 3),
            "target_hp": round(target["hp"], 3),
            "evaded": bool(damage_result.get("evaded")),
            "blocked": bool(damage_result.get("blocked")),
            "mp_cost": round(mp_cost, 3),
            "sp_cost": round(sp_cost, 3),
            "effects": applied_effects,
        }
    )

    if target["hp"] <= 0 and target["alive"]:
        _record_enemy_death(session, target, logs)


def _process_enemy_attack(session: dict[str, Any], enemy: dict[str, Any], logs: list[dict[str, Any]]) -> None:
    if not enemy["alive"] or not session["player"]["alive"]:
        return

    player = session["player"]
    skill = _choose_enemy_skill(session, enemy, player)
    mp_cost = max(0.0, _to_float(skill.get("mp_cost"), 0.0))
    sp_cost = max(0.0, _to_float(skill.get("sp_cost"), 0.0))
    if mp_cost > 0:
        enemy["mp"] = _clamp(_to_float(enemy.get("mp"), 0.0) - mp_cost, 0.0, _to_float(enemy.get("max_mp"), 0.0))
    if sp_cost > 0:
        enemy["sp"] = _clamp(_to_float(enemy.get("sp"), 0.0) - sp_cost, 0.0, _to_float(enemy.get("max_sp"), 0.0))

    applied_effects = _apply_skill_effects(session, skill, enemy, player)

    damage_result = {
        "damage": 0.0,
        "raw_damage": 0.0,
        "blocked_reduction": 0.0,
        "evaded": False,
        "blocked": False,
    }
    dmg = 0.0
    raw_damage = 0.0
    blocked_reduction = 0.0
    evaded = False
    if _skill_has_damage(skill):
        damage_result = _calc_damage_result(enemy["stats"], player["stats"], skill)
        dmg = _to_float(damage_result.get("damage"), 0.0)
        raw_damage = _to_float(damage_result.get("raw_damage"), dmg)
        blocked_reduction = _to_float(damage_result.get("blocked_reduction"), 0.0)
        evaded = bool(damage_result.get("evaded"))
        player["hp"] = _clamp(player["hp"] - dmg, 0.0, player["max_hp"])

    cast_time = max(0.1, _to_float(skill.get("cast_time"), _to_float(enemy["stats"].get("attack_interval"), 2.0)))
    cooldown = max(0.0, _to_float(skill.get("cooldown"), 0.0))
    enemy["next_ready_time"] = session["time"] + cast_time
    enemy["last_action_duration"] = cast_time
    enemy["last_skill_id"] = skill.get("id")
    enemy["last_skill_name"] = skill.get("name") or skill.get("id")
    if not isinstance(enemy.get("cooldowns"), dict):
        enemy["cooldowns"] = {}
    enemy["cooldowns"][skill["id"]] = session["time"] + cooldown

    if evaded:
        _add_pending_skill_exp(session, "ranging", 2.0 * max(0.0, raw_damage))
    else:
        if dmg > 0:
            _add_pending_skill_exp(session, "resilience", 1.5 * dmg)
            _add_pending_skill_exp(session, "defense", 0.5 * dmg)
        if blocked_reduction > 0:
            _add_pending_skill_exp(session, "defense", 2.0 * blocked_reduction)

    logs.append(
        {
            "type": "enemy_attack",
            "enemy_id": enemy["enemy_id"],
            "enemy_instance_id": enemy["instance_id"],
            "enemy_name": enemy["name"],
            "skill_id": skill.get("id"),
            "skill_name": skill.get("name") or skill.get("id"),
            "damage": round(dmg, 3),
            "raw_damage": round(raw_damage, 3),
            "blocked_reduction": round(blocked_reduction, 3),
            "evaded": evaded,
            "blocked": bool(damage_result.get("blocked")),
            "player_hp": round(player["hp"], 3),
            "effects": applied_effects,
            "mp_cost": round(mp_cost, 3),
            "sp_cost": round(sp_cost, 3),
        }
    )

    if player["hp"] <= 0 and player["alive"]:
        _handle_player_down(session, logs)

File: backend/internal/battle/combat.go
package battle

import (
	"math"

	"github.com/Milkyway-Liluo-Idle-Develop-Team/Milkyway-Liluo-Idle-2/backend/internal/attribute"
	"github.com/Milkyway-Liluo-Idle-Develop-Team/Milkyway-Liluo-Idle-2/backend/internal/gameconfig"
)

// Fallback skill IDs for runtime-generated basic attacks.
// Negative values avoid collision with registry-assigned positive IDs.
const (
	FallbackBasicAttackID      gameconfig.BattleSkillID = -1
	FallbackEnemyBasicAttackID gameconfig.BattleSkillID = -2
)

// processPlayerAttack resolves a single player's next attack.
func (s *BattleSession) processPlayerAttack(player *PlayerBattleEntity, target *EnemyBattleEntity) []BattleLog {
	skill := s.choosePlayerSkill(player, target)
	return s.resolveAttack(player, target, skill)
}

// processEnemyAttack resolves a single enemy's attack on its chosen target.
func (s *BattleSession) processEnemyAttack(enemy *EnemyBattleEntity, target *PlayerBattleEntity) []BattleLog {
	skill := s.chooseEnemySkill(enemy, target)
	return s.resolveAttack(enemy, target, skill)
}

// resolveAttack handles the full attack flow: costs, effects, damage, XP, death.
func (s *BattleSession) resolveAttack(attacker, defender BattleEntity, skill *BattleSkill) []BattleLog {
	var logs []BattleLog

	// Deduct costs.
	mpCost := max(0.0, skill.MPCost)
	spCost := max(0.0, skill.SPCost)
	if mpCost > 0 {
		attacker.SetMP(clamp(attacker.MP()-mpCost, 0.0, attacker.MaxMP()))
	}
	if spCost > 0 {
		attacker.SetSP(clamp(attacker.SP()-spCost, 0.0, attacker.MaxSP()))
	}

	// Apply skill effects.
	appliedEffects := s.applySkillEffects(attacker, defender, skill)

	// Calculate damage.
	result := CalcDamage(attacker, defender, skill, s.rng)

	// Apply damage.
	if result.Damage > 0 {
		defender.SetHP(clamp(defender.HP()-result.Damage, 0.0, defender.MaxHP()))
	}

	// Update hate: player dealing damage to enemy generates hate on that enemy.
	if attacker.Team() == TeamPlayer && defender.Team() == TeamEnemy && result.Damage > 0 {
		s.addHate(defender.EntityID(), attacker.EntityID(), result.Damage)
	}

	// Set attacker timing.
	castTime := skill.CastTime
	if castTime <= 0 {
		castTime = max(0.1, attacker.GetFinal(AttrAttackInterval))
	}
	cooldown := max(0.0, skill.Cooldown)

	attacker.SetNextReadyTime(s.Time + castTime)
	attacker.SetLastActionDuration(castTime)
	attacker.SetLastSkillID(skill.ID)
	if cooldown > 0 {
		attacker.SetCooldown(skill.ID, s.Time+cooldown)
	}

	// Build log.
	var logType BattleLogType
	if attacker.Team() == TeamPlayer {
		logType = BattleLogTypePlayerAttack
	} else {
		logType = BattleLogTypeEnemyAttack
	}

	log := BattleLog{
		Type:             logType,
		SkillID:          skill.ID,
		Damage:           result.Damage,
		RawDamage:        result.RawDamage,
		Evaded:           result.Evaded,
		Blocked:          result.Blocked,
		BlockedReduction: result.BlockedReduction,
		MPCost:           mpCost,
		SPCost:           spCost,
		Effects:          appliedEffects,
		AttackerEntityID: attacker.EntityID(),
		DefenderEntityID: defender.EntityID(),
		DefenderHP:       round3(defender.HP()),
	}
	logs = append(logs, log)

	// Skill XP (1 XP per skill use) for player attacks.
	if attacker.Team() == TeamPlayer && skill.ID > 0 {
		s.addPendingBattleSkillXP(skill.ID, 1.0)
	}

	// Combat XP from damage dealt.
	if attacker.Team() == TeamPlayer && result.Damage > 0 {
		s.awardCombatXP(result, skill)
	}

	// Combat XP from damage received (evade / block / hit).
	if defender.Team() == TeamPlayer {
		s.awardDefenseXP(result)
	}

	// Check death.
	if defender.HP() <= 0 && defender.Alive() {
		defender.SetAlive(false)
		if defender.Team() == TeamEnemy {
			logs = append(logs, s.handleEnemyDeath(defender.(*EnemyBattleEntity))...)
		} else {
			logs = append(logs, s.handlePlayerDown(defender.(*PlayerBattleEntity))...)
		}
	}

	return logs
}

// choosePlayerSkill selects the highest-priority usable skill for a specific player.
func (s *BattleSession) choosePlayerSkill(player *PlayerBattleEntity, target *EnemyBattleEntity) *BattleSkill {
	aliveEnemies := s.AliveEnemies()

	for _, entry := range player.SkillPlan() {
		skill, ok := player.Skills()[entry.SkillID]
		if !ok {
			continue
		}
		if !s.canUseSkill(player, skill) {
			continue
		}
		if entry.Condition != nil && !evaluateCondition(entry.Condition, player, target, toEntities(aliveEnemies)) {
			continue
		}
		if !skillHasDamage(skill) && skillRequiresEffectChange(skill) && !skillWouldApplyEffect(skill, player, target, s.Time) {
			continue
		}
		return skill
	}

	// Fallback to basic attack.
	if basic, ok := player.Skills()[player.BasicSkillID()]; ok {
		return basic
	}
	// Ultimate fallback.
	return &BattleSkill{
		ID:   FallbackBasicAttackID,
		Name: "基础攻击",
		Damage: &DamageProfile{
			Type:       "physical",
			Flat:       0,
			Multiplier: 1.0,
		},
		CastTime: max(0.1, player.GetFinal(AttrAttackInterval)),
	}
}

// chooseEnemySkill selects the highest-priority usable enemy skill.
func (s *BattleSession) chooseEnemySkill(enemy *EnemyBattleEntity, target BattleEntity) *BattleSkill {
	aliveAllies := s.AliveEnemies()

	for _, entry := range enemy.SkillPlan() {
		skill, ok := enemy.Skills()[entry.SkillID]
		if !ok {
			continue
		}
		if !s.canUseSkill(enemy, skill) {
			continue
		}
		if entry.Condition != nil && !evaluateCondition(entry.Condition, enemy, target, toEntities(aliveAllies)) {
			continue
		}
		if !skillHasDamage(skill) && skillRequiresEffectChange(skill) && !skillWouldApplyEffect(skill, enemy, target, s.Time) {
			continue
		}
		return skill
	}

	if basic, ok := enemy.Skills()[enemy.BasicSkillID()]; ok {
		return basic
	}
	return &BattleSkill{
		ID:   FallbackEnemyBasicAttackID,
		Name: "基础攻击",
		Damage: &DamageProfile{
			Type:       "physical",
			Flat:       0,
			Multiplier: 1.0,
		},
		CastTime: max(0.1, enemy.GetFinal(AttrAttackInterval)),
	}
}

func (s *BattleSession) canUseSkill(e BattleEntity, skill *BattleSkill) bool {
	if s.Time < e.Cooldowns()[skill.ID] {
		return false
	}
	if e.MP() < skill.MPCost {
		return false
	}
	if e.SP() < skill.SPCost {
		return false
	}
	return true
}

func skillHasDamage(skill *BattleSkill) bool {
	return skill.Damage != nil && (skill.Damage.Flat != 0 || skill.Damage.Multiplier != 0)
}

func skillRequiresEffectChange(skill *BattleSkill) bool {
	return len(skill.Effects) > 0
}

func skillWouldApplyEffect(skill *BattleSkill, caster, target BattleEntity, now float64) bool {
	for _, eff := range skill.Effects {
		dst := caster
		if eff.Target == "target" {
			dst = target
		}
		if dst == nil {
			continue
		}

		// Build the effect key as ApplyEffect would.
		candidate := ActiveEffect{
			SourceSkillID: skill.ID,
			Attribute:     eff.Attribute,
			Mode:          eff.Mode,
		}
		key := candidate.effectKey()

		found := false
		for _, ae := range dst.ActiveEffects() {
			if ae.effectKey() != key {
				continue
			}
			found = true
			// If the existing effect has expired, it will be cleaned up soon,
			// so re-applying makes sense.
			if ae.ExpiresAt != nil && *ae.ExpiresAt <= now {
				return true
			}
			// If the value differs (stronger buff, weaker debuff, etc.),
			// applying it would change the entity's stats.
			if ae.Value != eff.Value {
				return true
			}
			// Same value, still alive → this particular effect would change nothing.
		}
		if !found {
			// No existing effect with this key → applying creates a new modifier.
			return true
		}
	}
	return false
}

func (s *BattleSession) applySkillEffects(caster, target BattleEntity, skill *BattleSkill) []AppliedEffect {
	var applied []AppliedEffect
	for _, eff := range skill.Effects {
		dst := caster
		if eff.Target == "target" {
			dst = target
		}
		if dst == nil {
			continue
		}

		var expires *float64
		if eff.Duration > 0 {
			t := s.Time + eff.Duration
			expires = &t
		}

		modeStr := "flat"
		mode := EffectModeFlat
		if eff.Mode == EffectModePercentMult {
			mode = EffectModePercentMult
			modeStr = "percent_mult"
		}

		dst.ApplyEffect(ActiveEffect{
			SourceSkillID: skill.ID,
			Attribute:     eff.Attribute,
			Mode:          mode,
			Value:         eff.Value,
			ExpiresAt:     expires,
		}, s.Time)

		attrName, _ := attribute.Get().AttrString(eff.Attribute)
		applied = append(applied, AppliedEffect{
			Target:    eff.Target,
			Attribute: attrName,
			Mode:      modeStr,
			Value:     eff.Value,
			Duration:  eff.Duration,
		})
	}
	return applied
}

func (s *BattleSession) awardCombatXP(result DamageResult, skill *BattleSkill) {
	if result.Damage <= 0 {
		return
	}
	dmg := result.Damage
	if skill.Damage != nil && skill.Damage.Type == "magic" {
		if id, ok := gameconfig.StringToSkillID("magic"); ok {
			s.addPendingSkillXP(id, 1.0*dmg)
		}
		return
	}

	style := skill.PhysicalStyle
	if style == "" {
		style = "melee"
	}
	if style == "ranged" {
		if id, ok := gameconfig.StringToSkillID("strength"); ok {
			s.addPendingSkillXP(id, 0.5*dmg)
		}
		if id, ok := gameconfig.StringToSkillID("ranging"); ok {
			s.addPendingSkillXP(id, 0.5*dmg)
		}
	} else {
		if id, ok := gameconfig.StringToSkillID("strength"); ok {
			s.addPendingSkillXP(id, 0.8*dmg)
		}
		if id, ok := gameconfig.StringToSkillID("ranging"); ok {
			s.addPendingSkillXP(id, 0.2*dmg)
		}
	}
}

func (s *BattleSession) awardDefenseXP(result DamageResult) {
	if result.Evaded {
		if id, ok := gameconfig.StringToSkillID("ranging"); ok {
			s.addPendingSkillXP(id, 2.0*result.RawDamage)
		}
		return
	}
	if result.Damage > 0 {
		if id, ok := gameconfig.StringToSkillID("resilience"); ok {
			s.addPendingSkillXP(id, 1.5*result.Damage)
		}
		if id, ok := gameconfig.StringToSkillID("defense"); ok {
			s.addPendingSkillXP(id, 0.5*result.Damage)
		}
	}
	if result.BlockedReduction > 0 {
		if id, ok := gameconfig.StringToSkillID("defense"); ok {
			s.addPendingSkillXP(id, 2.0*result.BlockedReduction)
		}
	}
}

func (s *BattleSession) addPendingBattleSkillXP(skillID gameconfig.BattleSkillID, xp float64) {
	if xp <= 0 {
		return
	}
	s.PendingBattleSkillExp[skillID] += xp
}

func (s *BattleSession) addPendingSkillXP(skillID gameconfig.SkillID, xp float64) {
	if xp <= 0 {
		return
	}
	s.PendingSkillExp[skillID] += xp
}

func (s *BattleSession) handleEnemyDeath(enemy *EnemyBattleEntity) []BattleLog {
	return []BattleLog{{
		Type:             BattleLogTypeEnemyDied,
		AttackerEntityID: enemy.EntityID(),
		WaveNumber:       s.WaveNumber,
	}}
}

func (s *BattleSession) handlePlayerDown(player *PlayerBattleEntity) []BattleLog {
	respawn := s.Time + max(0.1, s.Config.Interval)
	s.RespawnTimes[player.EntityID()] = &respawn
	return []BattleLog{{
		Type:             BattleLogTypePlayerDowned,
		DefenderEntityID: player.EntityID(),
		DefenderHP:       0,
		NextWaveIn:       s.Config.Interval,
	}}
}

// addHate accumulates hate for a player against a specific enemy.
func (s *BattleSession) addHate(enemyEntityID, playerEntityID int64, amount float64) {
	if s.HateMap[enemyEntityID] == nil {
		s.HateMap[enemyEntityID] = make(map[int64]float64)
	}
	s.HateMap[enemyEntityID][playerEntityID] += amount
}

func round3(v float64) float64 {
	return math.Round(v*1000) / 1000
}

// evaluateCondition evaluates a skill condition against the current battle state.
// Stub: full implementation in a follow-up step.
func toEntities(enemies []*EnemyBattleEntity) []BattleEntity {
	out := make([]BattleEntity, len(enemies))
	for i, e := range enemies {
		out[i] = e
	}
	return out
}

func evaluateCondition(cond *SkillCondition, player, target BattleEntity, aliveEnemies []BattleEntity) bool {
	if cond == nil {
		return true
	}
	// TODO: implement full condition evaluation.
	return true
}
